server/util/clickhouse/clickhouse.go

```diff
@@ -259,6 +255,112 @@ func (e *Execution) AdditionalFields() []string {
 	}
 }
 
+// TestTargetStatus represents the status of a target, the target info and invocation details
+type TestTargetStatus struct {
+	GroupID       string
+	Role          string
+	RepoURL       string
+	Label         string
+	CommitSHA     string
+	CreatedAtUsec int64
+
+	RuleType       string
+	UserID         string
+	InvocationUUID string
+	TargetType     int32
+	TestSize       int32
+	Status         int32
+	StartTimeUsec  int64
+	DurationUsec   int64
+	BranchName     string
+}
+
+func (t *TestTargetStatus) ExcludedFields() []string {
+	return []string{}
+}
+
+func (t *TestTargetStatus) AdditionalFields() []string {
+	return []string{}
+}
+
+func (t *TestTargetStatus) TableName() string {
+	return "TestTargetStatuses"
+}
+
+func (t *TestTargetStatus) TableOptions() string {
+	return fmt.Sprintf("ENGINE=%s ORDER BY (group_id, role, repo_url, label, commit_sha, created_at_usec)", getEngine())
+}
+
+func hasProjection(db *gorm.DB, table Table, projectionName string) (bool, error) {
+	currentDatabase := db.Migrator().CurrentDatabase()
+
+	showCreateTableSQL := fmt.Sprintf("SHOW CREATE TABLE %s.%s", currentDatabase, table.TableName())
+	var createStmt string
+	if err := db.Raw(showCreateTableSQL).Row().Scan(&createStmt); err != nil {
+		return false, err
+	}
+
+	projections := extractProjectionNamesFromCreateStmt(createStmt)
+
+	_, ok := projections[projectionName]
+
+	return ok, nil
+}
+
+func addProjectionIfNotExist(db *gorm.DB, table Table, projectionName string, query string) error {
+	hasProjection, err := hasProjection(db, table, projectionName)
+	if err != nil {
+		return status.InternalErrorf("failed to check whether projection %q exists: %s", projectionName, err)
+	}
+	if hasProjection {
+		return nil
+	}
+	projectionStmt := fmt.Sprintf("ALTER TABLE %s ADD PROJECTION %s (%s)", table.TableName(), projectionName, query)
+	return db.Exec(projectionStmt).Error
+}
+
+const (
+	beforeCreateBody int = iota
+	inCreateBody
+	inProjection
+	afterCreateBody
+)
+
+// adapted from https://github.com/go-gorm/clickhouse/blob/master/migrator.go
```

Comment:
nit: add a comment with the context for why we couldn't use GORM's migration code directly? (I guess because it doesn't support migrating PROJECTION ?)

Reply:
Done.